From 07f790094dfce6e20bc00d5fa1572953a1b3fdb4 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:43:22 -0400
Subject: [ethernet] cxgb4: Check if rx checksum offload is enabled, while reading hardware calculated checksum

Message-id: <1410475447-94392-64-git-send-email-svemuri@redhat.com>
Patchwork-id: 93856
O-Subject: [RHEL7.1 PATCH BZ 1124947 063/108] cxgb4: Check if rx checksum offload is enabled, while reading hardware calculated checksum
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Signed-off-by: Hariprasad Shenai <hariprasad@chelsio.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit cca2822d37f419280b8cca27a4a6220e47afb77f)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/chelsio/cxgb4/sge.c b/drivers/net/ethernet/chelsio/cxgb4/sge.c
index 1564b0a..34ec62d 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/sge.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/sge.c
@@ -1696,7 +1696,8 @@ int t4_ethrx_handler(struct sge_rspq *q, const __be64 *rsp,
   return handle_trace_pkt(q->adap, si);
 
  pkt = (const struct cpl_rx_pkt *)rsp;
- csum_ok = pkt->csum_calc && !pkt->err_vec;
+ csum_ok = pkt->csum_calc && !pkt->err_vec &&
+    (q->netdev->features & NETIF_F_RXCSUM);
  if ((pkt->l2info & htonl(RXF_TCP)) &&
      (q->netdev->features & NETIF_F_GRO) && csum_ok && !pkt->ip_frag) {
   do_gro(rxq, si, pkt);
@@ -1718,8 +1719,7 @@ int t4_ethrx_handler(struct sge_rspq *q, const __be64 *rsp,
 
  rxq->stats.pkts++;
 
- if (csum_ok && (q->netdev->features & NETIF_F_RXCSUM) &&
-     (pkt->l2info & htonl(RXF_UDP | RXF_TCP))) {
+ if (csum_ok && (pkt->l2info & htonl(RXF_UDP | RXF_TCP))) {
   if (!pkt->ip_frag) {
    skb->ip_summed = CHECKSUM_UNNECESSARY;
    rxq->stats.rx_cso++;
-- 
1.7.1